### PR TITLE
[Flux] Refactor flux config: Add FluxTokenizerContainer, remove Validation class

### PR DESCRIPTION
**Pull request by wwwjn**
- Add FluxTokenizerContainer: Enforce tokenizer field to be not None
- Separate tokenizer and encoder configs: Split the old Encoder class into FluxTokenizerContainer.Config (tokenizer paths) and FluxEncoderConfig (encoder paths + autoencoder)
- Renamed `classifier_free_guidance_prob` → `prompt_dropout_prob`: Avoids confusion with inference-time CFG settings (enable_classifier_free_guidance, classifier_free_guidance_scale)

### CODE REVIEW

**Comment by wwwjn** [comment]
@claude Review PR

File: torchtitan/models/flux/configs.py [hunk 1]
```diff
@@ -8,35 +8,19 @@
...
-class Validation:
-    enable_classifier_free_guidance: bool = False
-    """Whether to use classifier-free guidance during sampling"""
-    classifier_free_guidance_scale: float = 5.0
-    """Classifier-free guidance scale when sampling"""
-    denoising_steps: int = 50
-    """How many denoising steps to sample when generating an image"""
-    eval_freq: int = 100
-    """Frequency of evaluation/sampling during training"""
+    random_init: bool = False
```
Comment (tianyu-l) [inline torchtitan/models/flux/configs.py:22]:
```suggestion
    _random_init: bool = False
```
since it's for test only

File: torchtitan/models/flux/tokenizer.py [hunk 2]
```diff
@@ -8,12 +8,58 @@
...
+
+    def encode(self, *args, **kwargs) -> list[int]:
+        raise NotImplementedError(
+            "Use t5_tokenizer.encode() or clip_tokenizer.encode() directly"
+        )
+
+    def decode(self, *args, **kwargs) -> str:
+        raise NotImplementedError(
+            "Use t5_tokenizer.decode() or clip_tokenizer.decode() directly"
+        )
```
Comment (tianyu-l) [inline torchtitan/models/flux/tokenizer.py]: Let's implement encode/decode as wrapper functions over t5 and clip encode/decode to return a tuple / dict of results, so that the dataloader / model never needs to call t5 / clip tokenizers directly.

File: torchtitan/models/flux/flux_datasets.py [hunk 3]
```diff
@@ -378,27 +376,28 @@ class Config(ParallelAwareDataloader.Config):
...
-        encoder: Encoder = field(default_factory=Encoder)
-        """This is a hack to get the T5 and CLIP tokenizer asset paths. Ideally tokenizer should be
-        built from the trainer, not inside dataloader. The reason we are doing this is because FLUX
-        has two tokenizer instead of just one."""
-
-        hf_assets_path: str = "./tests/assets/tokenizer"
-        """Similar to above, this is a hack to get the test tokenizer asset paths."""
+        def __post_init__(self):
+            if self.generate_timesteps and self.prompt_dropout_prob != 0.0:
+                logger.warning(
```
Comment (tianyu-l) [inline torchtitan/models/flux/flux_datasets.py]: let's error out instead of print warning

File: torchtitan/models/flux/configs.py [hunk 4]
```diff
@@ -51,3 +35,9 @@ class Inference:
     """Batch size for inference"""
     img_size: int = 256
     """Image size for inference"""
+    enable_classifier_free_guidance: bool = False
```
Comment (tianyu-l) [inline torchtitan/models/flux/configs.py]: so we are putting these configs in both validator.config and inference.config?

File: torchtitan/models/flux/flux_datasets.py [hunk 5]
```diff
@@ -378,27 +376,28 @@ class Config(ParallelAwareDataloader.Config):
         infinite: bool = True
         """Whether to loop the dataset infinitely"""
 
-        # TODO: In validation it should always be 0.0. Find a way to enforce this.
-        classifier_free_guidance_prob: float = 0.0
-        """Classifier-free guidance with probability `p` to dropout each text encoding independently.
-        If `n` text encoders are used, the unconditional model is trained in `p ^ n` of all steps.
-        For example, if `n = 2` and `p = 0.447`, the unconditional model is trained in 20% of all steps"""
+        prompt_dropout_prob: float = 0.0
```
Comment (tianyu-l) [inline torchtitan/models/flux/flux_datasets.py:367]: How legit is this name change -- what does FLUX repo use?

**Comment by wwwjn** [comment]
@claude review

File: torchtitan/models/flux/configs.py (hunk 4 above)
Comment (wwwjn) [reply torchtitan/models/flux/configs.py]: Moved to a shared config as claude suggested, these duplicated fields are used for generating / sampling a image in both validation and inference. h

File: torchtitan/models/flux/flux_datasets.py (hunk 5 above)
Comment (wwwjn) [reply torchtitan/models/flux/flux_datasets.py:367]: Flux repo does have dataloader function, they only opensource the inference / sampling related code, so we don't have direct corresponding name for this field

File: torchtitan/models/flux/inference/infer.py [hunk 6]
```diff
@@ -38,9 +37,9 @@ def inference(config: FluxTrainer.Config):
     prompts = original_prompts[global_rank::world_size]
 
     trainer.checkpointer.load(step=config.checkpoint.load_step)
-    t5_tokenizer, clip_tokenizer = build_flux_tokenizer(
-        config.encoder, config.hf_assets_path
-    )
+
+    # Build tokenizers from the config
+    tokenizer_container = config.tokenizer.build()
```
Comment (tianyu-l) [inline torchtitan/models/flux/inference/infer.py]: nit: no need to call it container -- you actually wouldn't know from this file

File: torchtitan/models/flux/tokenizer.py [hunk 7]
```diff
@@ -8,12 +8,80 @@
...
+
+        Args:
+            text: A string or list of strings to encode.
+
+        Returns:
+            A dict with keys "clip_tokens" and "t5_tokens", each mapping to a torch.Tensor.
+        """
+        return {
+            "clip_tokens": self.clip_tokenizer.encode(text),
+            "t5_tokens": self.t5_tokenizer.encode(text),
```
Comment (tianyu-l) [inline torchtitan/models/flux/tokenizer.py]: Do you think we can just call them
```suggestion
            "clip": self.clip_tokenizer.encode(text),
            "t5": self.t5_tokenizer.encode(text),
```

File: torchtitan/models/flux/tokenizer.py [hunk 8]
```diff
@@ -8,12 +8,80 @@
...
+            tokens: A dict with keys "clip_tokens" and/or "t5_tokens".
+
+        Returns:
+            A dict with keys "clip_text" and/or "t5_text".
+        """
+        result = {}
+        if "t5_tokens" in tokens:
+            result["t5_text"] = self.t5_tokenizer.decode(tokens["t5_tokens"])
+        if "clip_tokens" in tokens:
+            result["clip_text"] = self.clip_tokenizer.decode(tokens["clip_tokens"])
```
Comment (tianyu-l) [inline torchtitan/models/flux/tokenizer.py]:
```suggestion
        if "t5" in tokens:
            result["t5"] = self.t5_tokenizer.decode(tokens["t5"])
        if "clip_tokens" in tokens:
            result["clip"] = self.clip_tokenizer.decode(tokens["clip"])
```

File: torchtitan/models/flux/tokenizer.py [hunk 9]
```diff
@@ -8,12 +8,80 @@
...
+    def decode(self, tokens: dict[str, list[int]]) -> dict[str, str]:
+        """Decode token IDs using both T5 and CLIP tokenizers.
+
+        Args:
+            tokens: A dict with keys "clip_tokens" and/or "t5_tokens".
+
+        Returns:
+            A dict with keys "clip_text" and/or "t5_text".
+        """
+        result = {}
```
Comment (tianyu-l) [inline torchtitan/models/flux/tokenizer.py]: nit
```suggestion
        results = {}
```

**Comment by wwwjn** [comment]
>  It's probably worth to have a loss test, because this diffusion model is using many different components than llm.

Here's loss comparison - it's identical before vs after

```
[LOSS_COMPARE] Step    Baseline Loss    Test Loss   Difference
[LOSS_COMPARE] ----    -------------    ---------   ----------
[LOSS_COMPARE] 1       1.8155529499053955    1.8155529499053955   0.000000
[LOSS_COMPARE] 2       1.8347270488739014    1.8347270488739014   0.000000
[LOSS_COMPARE] 3       1.6217622756958008    1.6217622756958008   0.000000
[LOSS_COMPARE] 4       1.5435457229614258    1.5435457229614258   0.000000
[LOSS_COMPARE] 5       1.6602725982666016    1.6602725982666016   0.000000
[LOSS_COMPARE] 6       1.495316982269287    1.495316982269287   0.000000
[LOSS_COMPARE] 7       1.7248778343200684    1.7248778343200684   0.000000
[LOSS_COMPARE] 8       1.6782728433609009    1.6782728433609009   0.000000
[LOSS_COMPARE] 9       1.9163594245910645    1.9163594245910645   0.000000
[LOSS_COMPARE] 10      1.8467049598693848    1.8467049598693848   0.000000
[LOSS_COMPARE] 11      1.4386065006256104    1.4386065006256104   0.000000
[LOSS_COMPARE] 12      1.806770920753479    1.806770920753479   0.000000
[LOSS_COMPARE] 13      1.4788243770599365    1.4788243770599365   0.000000
[LOSS_COMPARE] 14      1.4884754419326782    1.4884754419326782   0.000000
[LOSS_COMPARE] 15      1.4160133600234985    1.4160133600234985   0.000000
[LOSS_COMPARE] 16      1.4677839279174805    1.4677839279174805   0.000000
[LOSS_COMPARE] 17      1.391682744026184    1.391682744026184   0.000000
[LOSS_COMPARE] 18      1.4481796026229858    1.4481796026229858   0.000000
[LOSS_COMPARE] 19      1.3557770252227783    1.3557770252227783   0.000000
[LOSS_COMPARE] 20      1.5204499959945679    1.5204499959945679   0.000000
[LOSS_COMPARE] 21      1.362585186958313    1.362585186958313   0.000000
[LOSS_COMPARE] 22      1.6736187934875488    1.6736187934875488   0.000000
[LOSS_COMPARE] 23      1.344430685043335    1.344430685043335   0.000000
[LOSS_COMPARE] 24      1.459771990776062    1.459771990776062   0.000000
[LOSS_COMPARE] 25      1.3226020336151123    1.3226020336151123   0.000000
[LOSS_COMPARE] 26      1.3107167482376099    1.3107167482376099   0.000000
[LOSS_COMPARE] 27      1.310360312461853    1.310360312461853   0.000000
[LOSS_COMPARE] 28      1.345038890838623    1.345038890838623   0.000000
[LOSS_COMPARE] 29      1.3249025344848633    1.3249025344848633   0.000000
[LOSS_COMPARE] 30      1.26181960105896    1.26181960105896   0.000000
[LOSS_COMPARE] 31      1.2442164421081543    1.2442164421081543   0.000000
[LOSS_COMPARE] 32      1.2157622575759888    1.2157622575759888   0.000000
[LOSS_COMPARE] 33      1.1822305917739868    1.1822305917739868   0.000000
[LOSS_COMPARE] 34      1.318563461303711    1.318563461303711   0.000000
[LOSS_COMPARE] 35      1.2407960891723633    1.2407960891723633   0.000000
[LOSS_COMPARE] 36      1.2758138179779053    1.2758138179779053   0.000000
[LOSS_COMPARE] 37      1.1083598136901855    1.1083598136901855   0.000000
[LOSS_COMPARE] 38      1.223547339439392    1.223547339439392   0.000000
[LOSS_COMPARE] 39      1.1080725193023682    1.1080725193023682   0.000000
[LOSS_COMPARE] 40      1.3764539957046509    1.3764539957046509   0.000000
[LOSS_COMPARE] 41      1.2722375392913818    1.2722375392913818   0.000000
[LOSS_COMPARE] 42      1.2694342136383057    1.2694342136383057   0.000000
[LOSS_COMPARE] 43      1.1137546300888062    1.1137546300888062   0.000000
[LOSS_COMPARE] 44      1.2400405406951904    1.2400405406951904   0.000000
[LOSS_COMPARE] 45      1.045558214187622    1.045558214187622   0.000000
[LOSS_COMPARE] 46      1.1901164054870605    1.1901164054870605   0.000000
[LOSS_COMPARE] 47      1.1180517673492432    1.1180517673492432   0.000000
[LOSS_COMPARE] 48      1.1599459648132324    1.1599459648132324   0.000000
[LOSS_COMPARE] 49      1.0186576843261719    1.0186576843261719   0.000000
[LOSS_COMPARE] 50      1.1787950992584229    1.1787950992584229   0.000000
[LOSS_COMPARE] 51      1.033793568611145    1.033793568611145   0.000000
[LOSS_COMPARE] 52      1.0993987321853638    1.0993987321853638   0.000000
[LOSS_COMPARE] 53      0.9755196571350098    0.9755196571350098   0.000000
[LOSS_COMPARE] 54      1.0756466388702393    1.0756466388702393   0.000000
[LOSS_COMPARE] 55      1.060961365699768    1.060961365699768   0.000000
[LOSS_COMPARE] 56      1.188570499420166    1.188570499420166   0.000000
[LOSS_COMPARE] 57      1.0382428169250488    1.0382428169250488   0.000000
[LOSS_COMPARE] 58      1.2789831161499023    1.2789831161499023   0.000000
[LOSS_COMPARE] 59      0.8888002634048462    0.8888002634048462   0.000000
[LOSS_COMPARE] 60      1.125573992729187    1.125573992729187   0.000000
[LOSS_COMPARE] 61      1.0346546173095703    1.0346546173095703   0.000000
[LOSS_COMPARE] 62      1.1389732360839844    1.1389732360839844   0.000000
[LOSS_COMPARE] 63      1.0208430290222168    1.0208430290222168   0.000000
[LOSS_COMPARE] 64      1.0611464977264404    1.0611464977264404   0.000000
[LOSS_COMPARE] 65      1.0129194259643555    1.0129194259643555   0.000000
[LOSS_COMPARE] 66      1.0819472074508667    1.0819472074508667   0.000000
[LOSS_COMPARE] 67      0.901334285736084    0.901334285736084   0.000000
[LOSS_COMPARE] 68      1.1052348613739014    1.1052348613739014   0.000000
[LOSS_COMPARE] 69      0.9719484448432922    0.9719484448432922   0.000000
[LOSS_COMPARE] 70      0.9540094137191772    0.9540094137191772   0.000000
[LOSS_COMPARE] 71      0.9710623025894165    0.9710623025894165   0.000000
[LOSS_COMPARE] 72      1.0673062801361084    1.0673062801361084   0.000000
[LOSS_COMPARE] 73      1.1069374084472656    1.1069374084472656   0.000000
[LOSS_COMPARE] 74      1.0854978561401367    1.0854978561401367   0.000000
[LOSS_COMPARE] 75      0.932022213935852    0.932022213935852   0.000000
[LOSS_COMPARE] 76      1.0333012342453003    1.0333012342453003   0.000000
[LOSS_COMPARE] 77      1.0034865140914917    1.0034865140914917   0.000000
[LOSS_COMPARE] 78      1.0126731395721436    1.0126731395721436   0.000000
[LOSS_COMPARE] 79      0.9098399877548218    0.9098399877548218   0.000000
[LOSS_COMPARE] 80      1.0198639631271362    1.0198639631271362   0.000000
[LOSS_COMPARE] 81      0.9072495698928833    0.9072495698928833   0.000000
[LOSS_COMPARE] 82      1.0666074752807617    1.0666074752807617   0.000000
[LOSS_COMPARE] 83      0.9047322869300842    0.9047322869300842   0.000000
[LOSS_COMPARE] 84      0.9407524466514587    0.9407524466514587   0.000000
[LOSS_COMPARE] 85      0.9388478398323059    0.9388478398323059   0.000000
[LOSS_COMPARE] 86      1.0260812044143677    1.0260812044143677   0.000000
[LOSS_COMPARE] 87      1.068752646446228    1.068752646446228   0.000000
[LOSS_COMPARE] 88      0.874472975730896    0.874472975730896   0.000000
[LOSS_COMPARE] 89      0.9945414066314697    0.9945414066314697   0.000000
[LOSS_COMPARE] 90      1.0483378171920776    1.0483378171920776   0.000000
[LOSS_COMPARE] 91      0.887055516242981    0.887055516242981   0.000000
[LOSS_COMPARE] 92      0.9463194012641907    0.9463194012641907   0.000000
[LOSS_COMPARE] 93      0.9075696468353271    0.9075696468353271   0.000000
[LOSS_COMPARE] 94      1.0015450716018677    1.0015450716018677   0.000000
[LOSS_COMPARE] 95      0.9358677864074707    0.9358677864074707   0.000000
[LOSS_COMPARE] 96      1.0769689083099365    1.0769689083099365   0.000000
[LOSS_COMPARE] 97      0.8972340226173401    0.8972340226173401   0.000000
[LOSS_COMPARE] 98      1.0144946575164795    1.0144946575164795   0.000000
[LOSS_COMPARE] 99      1.070130705833435    1.070130705833435   0.000000
[LOSS_COMPARE] 100     0.9780427813529968    0.9780427813529968   0.000000
[LOSS_COMPARE]
```

File: torchtitan/models/flux/configs.py (hunk 1 above)
Comment (wwwjn) [reply torchtitan/models/flux/configs.py:22]: I reverted this rename , still keep using `random_init`. I found a leading underscore doesn't work with tyro parse from CLI (because tyro will change all _ to be `-`)

File: tests/integration_tests/flux.py [hunk 10]
```diff
@@ -53,7 +53,11 @@ def run_single_test(test_flavor: OverrideDefinitions, output_dir: str):
     dump_folder_arg = f"--dump_folder {output_dir}/{test_name}"
 
     # Random init encoder for offline testing
+<<<<<<< HEAD
     random_init_arg = "--tokenizer.test_mode --encoder._random-init"
+=======
+    random_init_arg = "--tokenizer.test_mode --encoder.random_init"
+>>>>>>> 36c59d2b (Fix missed _random_init rename in unit test CLI argument)
```
Comment (tianyu-l) [inline tests/integration_tests/flux.py]: need to resolve conflict
